!!! TASK: Introduce first level content repository projection

mhsdesign opened the pull request:
Replaces the quirky `ContentRepositoryReadModelAdapterInterface` with a new `ContentGraphReadModelInterface` that is a 1st level dependency to the `ContentRepository`.
This simplifies code flow and prevents CR instances without its main projection to be created.

## Note

This is a breaking change because it changes the configuration format for the `ContentRepositoryRegistry`  package from
```yaml
Neos:
  ContentRepositoryRegistry:
    presets:
      'default':
        # ...
        projections:
          'Neos.ContentRepository:ContentGraph':
            catchUpHooks: # ...
          'Neos.Neos:DocumentUriPathProjection': #...
```

to

```yaml
Neos:
  ContentRepositoryRegistry:
    presets:
      'default':
        # ...
        contentGraphProjection:
          catchUpHooks: #...
        projections:
          'Neos.Neos:DocumentUriPathProjection': #...
```

mhsdesign committed:
FEATURE: Refactor cr readmodel projection to be first level

A ContentRepository cannot exist without this projection

mhsdesign committed:
BUGFIX: Simplify `projectionState` to be more reliable

Previously if called with `ProjectionStateInterface::class` it would return the first state which will then not be callable again by its fqn. Now only fqn are allowed.
Initially the interface logic was needed for interchangeable cr readmodel projection states

mhsdesign committed:
TASK: Remove ContentGraphInterface::countNodes as the method is not bound to a workspace but cross content stream

Also since its internal and only used for the tests it a good to be hidden in the low level `ContentRepositoryReadModel` next to other low level methods

mhsdesign commented on Neos.ContentRepository.Core/Classes/Projection/ProjectionsAndCatchUpHooks.php:
```diff
@@ -13,7 +13,8 @@
      * @param array<class-string<ProjectionInterface<ProjectionStateInterface>>, CatchUpHookFactories> $catchUpHookFactoriesByProjectionClassName
      */
     public function __construct(
-        public Projections $projections,
+        public ContentRepositoryReadModelProjection $readModelProjection,
+        public Projections $additionalProjections,
         private array $catchUpHookFactoriesByProjectionClassName,
```
todo discuss naming of `        public ContentRepositoryReadModelProjection $readModelProjection,` and if we like this structure here in ProjectionAndCatchupHooks

also the construction of this in the factory is a litte hacky

bwaidelich commented on Neos.ContentRepository.Core/Classes/ContentRepositoryReadModel.php:
```diff
@@ -26,118 +24,32 @@
+ * @internal This is just an implementation detail to be implemented by the specific adapters
  */
-final class ContentRepositoryReadModel implements ProjectionStateInterface
+interface ContentRepositoryReadModel extends ProjectionStateInterface
```
Interfaces must have the `Interface` suffix

bwaidelich commented on Neos.ContentRepository.Core/Classes/ContentRepository.php:
```diff
@@ -69,6 +70,23 @@ final class ContentRepository
      */
     private array $projectionStateCache;
 
+    /**
+     * @var array<string, ContentGraphInterface> Runtime cache for {@see ContentGraphInterface} instances, indexed by their workspace name
+     */
+    private array $contentGraphInstancesByWorkspaceName = [];
+
+    /**
+     * @var array<string, Workspace> Runtime cache for {@see Workspace} instances, indexed by their name
+     */
+    private array $workspaceInstancesByName = [];
+
+    /**
+     * @var array<string, ContentStream> Runtime cache for {@see ContentStream} instances, indexed by their name
+     */
+    private array $contentStreamInstancesById = [];
```
Are those new runtime caches related? I would prefer to leave them for a separate PR

mhsdesign replied:
removed

bwaidelich commented on Neos.ContentRepository.TestSuite/Classes/Behavior/Features/Bootstrap/CRTestSuiteTrait.php:
```diff
@@ -154,7 +156,20 @@ public function workspaceHasStatus(string $rawWorkspaceName, string $status): vo
+                {
+                };
+            }
+        };
```
We could declare some custom Test CR Service that just exposes all `ContentRepositoryServiceFactoryDependencies` to make these less of a hack – but that can be done in a follow-up of course

bwaidelich reviewed (commented):
Nice one and it's a good idea to move the `countNodes()` hack to the higher level interface.
I already added a couple of comments even though this is just a draft yet

mhsdesign committed:
TASK: Ignore phpstan error

mhsdesign commented on Neos.ContentRepository.Core/Classes/Projection/ContentRepositoryReadModelProjection.php:
```diff
@@ -0,0 +1,16 @@
+ * @extends ProjectionInterface<ContentRepositoryReadModel>
+ * @api
+ */
+interface ContentRepositoryReadModelProjection extends ProjectionInterface
```
```suggestion
interface ContentRepositoryProjectionInterface extends ProjectionInterface
```

mhsdesign commented on Neos.ContentRepository.Core/Classes/ContentRepositoryReadModel.php:
```diff
@@ -25,65 +24,26 @@
+ * @internal This is just an implementation detail to be implemented by the specific adapters
  */
-final class ContentRepositoryReadModel implements ProjectionStateInterface
+interface ContentRepositoryReadModel extends ProjectionStateInterface
```
```suggestion
interface ContentRepositoryReadModelInterface extends ProjectionStateInterface
```

bwaidelich commented on Neos.ContentRepository.Core/Classes/CommandHandlingDependencies.php:
```diff
@@ -38,8 +38,10 @@ final class CommandHandlingDependencies
-    {
+    public function __construct(
+        private readonly ContentRepository $contentRepository,
+        private readonly ContentRepositoryReadModel $adapter
```
"adapter" is not very specific, what about:
```suggestion
        private readonly ContentRepositoryReadModel $contentRepositoryReadModel
```

bwaidelich commented on Neos.ContentRepository.Core/Classes/ContentRepository.php:
```diff
@@ -69,8 +70,9 @@ final class ContentRepository
     private array $projectionStateCache;
 
-    private CommandHandlingDependencies $commandHandlingDependencies;
+    private ContentRepositoryReadModel $adapter;
```
Same here:
```suggestion
    private ContentRepositoryReadModel $contentRepositoryReadModel;
```

bwaidelich commented on Neos.ContentRepository.Core/Classes/Projection/ProjectionsAndCatchUpHooks.php:
```diff
@@ -13,7 +13,8 @@
      */
     public function __construct(
-        public Projections $projections,
+        public ContentRepositoryReadModelProjection $readModelProjection,
```
I don't really like that distinction here.. In this case we should not make a difference between the individual projections IMO

bwaidelich reviewed (commented):
@mhsdesign This is a great start already, but I'm not yet super happy with the split in the CR core.
I have some rough ideas I would like to discuss with you at some point

mhsdesign committed:
TASK: Adjust namings

mhsdesign committed:
TASK: Introduce dedicated `contentRepositoryProjection` and factory

mhsdesign committed:
TASK: Simplify ProjectionsAndCatchUpHooks

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into task/cr-readmodel-first-level

mhsdesign committed:
TASK: Adjust namings (again xD)

The name `ContentRepositoryProjectionInterface` doesnt specify what makes it different to `ProjectionInterface` and indeed its a graph that has to be implemented by the adapter

mhsdesign committed:
TASK: Adjust also naming of ContentGraphReadModel

mhsdesign committed:
TASK: Move content graph projection interfaces to `Projection\ContentGraph`

mhsdesign commented on Neos.ContentRepositoryRegistry/Classes/ContentRepositoryRegistry.php:
```diff
@@ -242,29 +245,40 @@ private function buildPropertySerializer(ContentRepositoryId $contentRepositoryI
+    /**
+     * @param class-string<ProjectionFactoryInterface<ProjectionInterface<ProjectionStateInterface>>> $expectedProjectionFactory
+     */
+    private function registerProjection(mixed $projectionOptions, ProjectionsAndCatchUpHooksFactory $projectionsFactory, string $expectedProjectionFactory, string $projectionName, ContentRepositoryId $contentRepositoryId): void
```
a little verbose ... sadly

mhsdesign commented on Neos.ContentRepository.Core/Classes/Factory/ProjectionsAndCatchUpHooksFactory.php:
```diff
@@ -71,9 +73,17 @@ public function build(ProjectionFactoryDependencies $projectionFactoryDependenci
             );
             $catchUpHookFactoriesByProjectionClassName[$projection::class] = $catchUpHookFactories;
-            $projectionsArray[] = $projection;
+            if ($projection instanceof ContentGraphProjectionInterface) {
+                $contentGraphProjection = $projection;
```
little bit dirty but what solls :D

mhsdesign replied:
todo throw here if `$contentGraphProjection` already set!

bwaidelich commented on Neos.ContentGraph.DoctrineDbalAdapter/src/ContentGraphReadModelAdapter.php:
```diff
@@ -32,10 +32,9 @@
+ * @internal
  */
-final readonly class ContentRepositoryReadModelAdapter implements ContentRepositoryReadModelAdapterInterface
+final readonly class ContentGraphReadModelAdapter implements ContentGraphReadModelInterface
```
Just wondering: Why did you rename this?
That read model knows all about workspaces etc. and those are not really part of the content graph, are they?

kitsunet replied:
They are?! We moved them to the contentgraph projection, IMHO because they are inherently part of it now?

kitsunet replied:
I think you missed this yesterday then, sebastian and me suggested that this should all be ContentGraph instead of content repository.

bwaidelich commented on Neos.ContentRepository.Core/Classes/CommandHandlingDependencies.php:
```diff
@@ -82,8 +85,11 @@ public function getContentGraph(WorkspaceName $workspaceName): ContentGraphInter
+        $workspace = $this->contentGraphReadModel->findWorkspaceByName($workspaceName);
+        if ($workspace === null) {
+            throw WorkspaceDoesNotExist::butWasSupposedTo($workspaceName);
+        }
```
With the [CR Privilges PR](https://github.com/neos/neos-development-collection/pull/5298) the evaluation of permissions is centralized to the [ContentRepository](https://github.com/neos/neos-development-collection/blob/bf57af6d351550860be19f28b8318889fc9d2836/Neos.ContentRepository.Core/Classes/ContentRepository.php#L250-L268)
I _think_ it's fine (or even desired) that they are _not_ evaluated here, but I'm not 100% sure

bwaidelich commented on Neos.ContentRepository.Core/Classes/ContentRepository.php:
```diff
@@ -141,18 +143,24 @@ public function handle(CommandInterface $command): void
-                $this->projectionStateCache[$projectionStateClassName] = $projectionState;
-                return $projectionState;
-            }
+        if (in_array(ContentGraphReadModelInterface::class, class_implements($projectionStateClassName), true)) {
+            throw new \InvalidArgumentException(sprintf('Accessing the internal content repository projection state via %s(%s) is not allowed. Please use the API on the content repository instead.', __FUNCTION__, $projectionStateClassName), 1729338679);
```
nitpick: why not move this check to the top and keep the rest like before?

mhsdesign replied:
then it would be more expensive imo as we call `class_implements` always and here its just to improve the error message ;)

kitsunet replied:
as said above I would just allow it and thrash the check. it's a projectionstate after all, just one with _better_ API to reach it.

bwaidelich replied:
We will have to disallow this with the [CR Privileges](https://github.com/neos/neos-development-collection/pull/5298) because they enforce the constraints on the CR instance

kitsunet replied:
right, ok fair enough. Then this place is fine for me.

bwaidelich commented on Neos.ContentRepository.Core/Classes/ContentRepository.php:
```diff
@@ -141,18 +143,24 @@ public function handle(CommandInterface $command): void
      */
     public function projectionState(string $projectionStateClassName): ProjectionStateInterface
     {
+        if (!isset($this->projectionStateCache)) {
+            foreach ($this->projectionsAndCatchUpHooks->projections as $projection) {
+                if ($projection instanceof ContentGraphProjectionInterface) {
+                    continue;
+                }
+                $projectionState = $projection->getState();
+                $this->projectionStateCache[$projectionState::class] = $projectionState;
+            }
+        }
```
I don't completely get the implications of this, but why do we build up the runtime cache for all projections once any state is retrieved?

mhsdesign replied:
because we did so previously almost as well just that we break on the first match, by building it once its more easy to grasp

bwaidelich commented on Neos.ContentRepository.Core/Classes/Projection/ProjectionStatuses.php:
```diff
@@ -18,9 +18,12 @@ private function __construct(
+    /**
+     * @param array<class-string<ProjectionInterface<ProjectionStateInterface>>, ProjectionStatus> $statuses
+     */
+    public static function create(array $statuses = []): self
```
nitpick: default values in method signatures can be error prone, I'd slighly prefer explicit constructors like
```php
public static function createEmpty(): self {...}
public static function fromArray(array $statuses): self {...}
```

bwaidelich reviewed (commented):
First set of comments, looking great already.
What puzzles me is the distinction between content*Repository*Projection  (in configuration etc) but content*Graph*ReadModel

kitsunet commented on Neos.ContentRepository.Core/Classes/ContentRepository.php:
```diff
@@ -141,18 +143,24 @@ public function handle(CommandInterface $command): void
     {
+        if (!isset($this->projectionStateCache)) {
+            foreach ($this->projectionsAndCatchUpHooks->projections as $projection) {
+                if ($projection instanceof ContentGraphProjectionInterface) {
```
So but that means you cannot get the CG projection here anymore? not sure that is good?!

bwaidelich replied:
I think that's a _good_ thing (see comment above re PR privileges)

kitsunet commented on Neos.ContentRepository.TestSuite/Classes/Behavior/Features/Bootstrap/CRTestSuiteTrait.php:
```diff
@@ -155,7 +157,20 @@ public function workspaceHasStatus(string $rawWorkspaceName, string $status): vo
     public function iExpectTheGraphProjectionToConsistOfExactlyNodes(int $expectedNumberOfNodes): void
     {
-        $actualNumberOfNodes = $this->currentContentRepository->getContentGraph($this->currentWorkspaceName)->countNodes();
+        // HACK to access
```
YIKES

kitsunet commented on Neos.ContentRepositoryRegistry/Classes/ContentRepositoryRegistry.php:
```diff
@@ -242,29 +245,40 @@ private function buildPropertySerializer(ContentRepositoryId $contentRepositoryI
+     */
+    private function registerProjection(mixed $projectionOptions, ProjectionsAndCatchUpHooksFactory $projectionsFactory, string $expectedProjectionFactory, string $projectionName, ContentRepositoryId $contentRepositoryId): void
+    {
+        (is_array($projectionOptions)) || throw InvalidConfigurationException::fromMessage('Projection "%s" (content repository "%s") must be configured as array got %s', $projectionName, $contentRepositoryId->value, get_debug_type($projectionOptions));
```
`(is_array($projectionOptions)) || throw InvalidConfigu...`

YIKES

also why is it mixed in the signature then

mhsdesign replied:
later alligator

kitsunet commented on Neos.ContentRepositoryRegistry/Classes/ContentRepositoryRegistry.php:
```diff
@@ -242,29 +245,40 @@ private function buildPropertySerializer(ContentRepositoryId $contentRepositoryI
+    {
+        (is_array($projectionOptions)) || throw InvalidConfigurationException::fromMessage('Projection "%s" (content repository "%s") must be configured as array got %s', $projectionName, $contentRepositoryId->value, get_debug_type($projectionOptions));
+        $projectionFactory = isset($projectionOptions['factoryObjectName']) ? $this->objectManager->get($projectionOptions['factoryObjectName']) : null;
+        if (!$projectionFactory instanceof $expectedProjectionFactory) {
```
Why do we need this here at all, aka why  do we need $expectedProjectionFactory?

mhsdesign replied:
refactored

mhsdesign committed:
TASK: Rename `contentRepositoryProjection` to `contentGraphProjection`

mhsdesign committed:
TASK: Simplify cr registry projection building and rename factory

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into task/cr-readmodel-first-level

bwaidelich committed:
Merge branch '9.0' into task/cr-readmodel-first-level

bwaidelich committed:
Merge branch '9.0' into task/cr-readmodel-first-level
